[valid-expect-in-promise] Another false positive

Hello, it's me again 😆. Thank you for the extremely quick fix to #930.

There are other cases where the rule is erroring even though the promise is (indirectly) awaited:

```js
it("throws then resolves", async () => {
  const p = Promise.reject(new Error("myError")).then(undefined, (e) => {
    expect(e).toBeInstanceOf(Error);

    return "handled";
  });

  await expect(p).resolves.toBe("handled");
});

it("throws twice", async () => {
  const p = Promise.reject(new Error("myError")).then(undefined, (e) => {
    expect(e).toBeInstanceOf(Error);

    throw new Error("myError2");
  });

  await expect(p).rejects.toThrow("myError2");
});
```

```text
  2:9  error  This promise should either be returned or awaited to ensure the expects in it's chain are called  jest/valid-expect-in-promise
  12:9  error  This promise should either be returned or awaited to ensure the expects in it's chain are called  jest/valid-expect-in-promise
```

**[Repository here](https://github.com/srmagura/eslint-plugin-jest-promise-repro2)**

### Versions

- "eslint": "^7.32.0",
- "eslint-plugin-jest": "^25.0.5"
- Node 14

--- a/src/rules/__tests__/valid-expect-in-promise.test.ts
+++ b/src/rules/__tests__/valid-expect-in-promise.test.ts
@@ -16,6 +16,50 @@ ruleTester.run('valid-expect-in-promise', rule, {
     'Promise.resolve().then(() => expect(1).toBe(2))',
     'const x = Promise.resolve().then(() => expect(1).toBe(2))',
     dedent`
+      it('is valid', () => {
+        const promise = loadNumber().then(number => {
+          expect(typeof number).toBe('number');
+
+          return number + 1;
+        });
+
+        expect(promise).resolves.toBe(1);
+      });
+    `,
+    dedent`
+      it('is valid', () => {
+        const promise = loadNumber().then(number => {
+          expect(typeof number).toBe('number');
+
+          return number + 1;
+        });
+
+        expect(promise).resolves.not.toBe(2);
+      });
+    `,
+    dedent`
+      it('is valid', () => {
+        const promise = loadNumber().then(number => {
+          expect(typeof number).toBe('number');
+
+          return number + 1;
+        });
+
+        expect(promise).rejects.toBe(1);
+      });
+    `,
+    dedent`
+      it('is valid', () => {
+        const promise = loadNumber().then(number => {
+          expect(typeof number).toBe('number');
+
+          return number + 1;
+        });
+
+        expect(promise).rejects.not.toBe(2);
+      });
+    `,
+    dedent`
       it('is valid', async () => {
         const promise = loadNumber().then(number => {
           expect(typeof number).toBe('number');
@@ -1459,6 +1503,46 @@ ruleTester.run('valid-expect-in-promise', rule, {
           });
 
           log(promise);
+        });
+      `,
+      errors: [
+        {
+          messageId: 'expectInFloatingPromise',
+          line: 2,
+          column: 9,
+        },
+      ],
+    },
+    {
+      code: dedent`
+        it('is valid', async () => {
+          const promise = loadNumber().then(number => {
+            expect(typeof number).toBe('number');
+
+            return number + 1;
+          });
+
+          expect(promise).toBeInstanceOf(Promise);
+        });
+      `,
+      errors: [
+        {
+          messageId: 'expectInFloatingPromise',
+          line: 2,
+          column: 9,
+        },
+      ],
+    },
+    {
+      code: dedent`
+        it('is valid', async () => {
+          const promise = loadNumber().then(number => {
+            expect(typeof number).toBe('number');
+
+            return number + 1;
+          });
+
+          expect(anotherPromise).resolves.toBe(1);
         });
       `,
       errors: [

--- a/src/rules/valid-expect-in-promise.ts
+++ b/src/rules/valid-expect-in-promise.ts
@@ -4,6 +4,7 @@ import {
 } from '@typescript-eslint/experimental-utils';
 import {
   KnownCallExpression,
+  ModifierName,
   createRule,
   getAccessorValue,
   getNodeName,
@@ -12,6 +13,7 @@ import {
   isIdentifier,
   isSupportedAccessor,
   isTestCaseCall,
+  parseExpectCall,
 } from './utils';
 
 type PromiseChainCallExpression = KnownCallExpression<
@@ -175,6 +177,28 @@ const isValueAwaitedInArguments = (
   return false;
 };
 
+const getLeftMostCallExpression = (
+  call: TSESTree.CallExpression,
+): TSESTree.CallExpression => {
+  let leftMostCallExpression: TSESTree.CallExpression = call;
+  let node: TSESTree.Node = call;
+
+  while (node) {
+    if (node.type === AST_NODE_TYPES.CallExpression) {
+      leftMostCallExpression = node;
+      node = node.callee;
+    }
+
+    if (node.type !== AST_NODE_TYPES.MemberExpression) {
+      break;
+    }
+
+    node = node.object;
+  }
+
+  return leftMostCallExpression;
+};
+
 /**
  * Attempts to determine if the runtime value represented by the given `identifier`
  * is `await`ed or `return`ed within the given `body` of statements
@@ -198,11 +222,27 @@ const isValueAwaitedOrReturned = (
 
     if (node.type === AST_NODE_TYPES.ExpressionStatement) {
       // it's possible that we're awaiting the value as an argument
-      if (
-        node.expression.type === AST_NODE_TYPES.CallExpression &&
-        isValueAwaitedInArguments(name, node.expression)
-      ) {
-        return true;
+      if (node.expression.type === AST_NODE_TYPES.CallExpression) {
+        if (isValueAwaitedInArguments(name, node.expression)) {
+          return true;
+        }
+
+        const leftMostCall = getLeftMostCallExpression(node.expression);
+
+        if (
+          isExpectCall(leftMostCall) &&
+          leftMostCall.arguments.length > 0 &&
+          isIdentifier(leftMostCall.arguments[0], name)
+        ) {
+          const { modifier } = parseExpectCall(leftMostCall);
+
+          if (
+            modifier?.name === ModifierName.resolves ||
+            modifier?.name === ModifierName.rejects
+          ) {
+            return true;
+          }
+        }
       }
 
       if (

